wasm-text.md

```diff
@@ -0,0 +1,278 @@
+
+```k
+    syntax FoldedInstr ::= "(" "block" OptionalId TypeDecls Instrs ")"
+ // ------------------------------------------------------------------
+    rule <k> ( block               TDECLS:TypeDecls INSTRS:Instrs ) => block    TDECLS INSTRS end ... </k>
+    rule <k> ( block ID:Identifier TDECLS:TypeDecls INSTRS:Instrs ) => block ID TDECLS INSTRS end ... </k>
+
+    syntax FoldedInstr ::= "(" "if" OptionalId TypeDecls Instrs "(" "then" Instrs ")" ")"
+                         | "(" "if" OptionalId TypeDecls Instrs "(" "then" Instrs ")" "(" "else" Instrs ")" ")"
+ // -----------------------------------------------------------------------------------------------------------
+    rule <k> ( if OID:OptionalId TDECLS:TypeDecls C:Instrs ( then IS ) )              => C ~> if OID TDECLS IS          end ... </k>
+    rule <k> ( if                TDECLS:TypeDecls C:Instrs ( then IS ) ( else IS' ) ) => C ~> if     TDECLS IS else IS' end ... </k>
+    rule <k> ( if  ID:Identifier TDECLS:TypeDecls C:Instrs ( then IS ) ( else IS' ) ) => C ~> if  ID TDECLS IS else IS' end ... </k>
+
+    syntax FoldedInstr ::= "(" "loop" OptionalId TypeDecls Instrs ")"
+ // -----------------------------------------------------------------
+    rule <k> ( loop               TDECLS:TypeDecls IS ) => loop    TDECLS IS end ... </k>
+    rule <k> ( loop ID:Identifier TDECLS:TypeDecls IS ) => loop ID TDECLS IS end ... </k>
+```
+
+Looking up Indices
+------------------
+
+In the abstract Wasm syntax, indices are always integers.
+In the text format, we extend indices to incorporate identifiers.
+
+```k
+    syntax Index ::= Identifier
+ // ---------------------------
+```
+
+We enable context lookups with identifiers.
+
+```k
+    rule #ContextLookup(IDS:Map, ID:Identifier) => {IDS [ ID ]}:>Int
+      requires ID in_keys(IDS)
+```
+
+Block Instructions
+------------------
+
+In the text format, block instructions can have identifiers attached to them, and branch instructions can refer to these identifiers.
+First, we allow identifiers on labels.
+
+```k
+    syntax Label ::= "label" Identifier VecType "{" Instrs "}" ValStack
+ // --------------------------------------------------------------------
+    rule <k> label _:Identifier [ TYPES ] { _ } VALSTACK' => . ... </k>
+         <valstack> VALSTACK => #take(TYPES, VALSTACK) ++ VALSTACK' </valstack>
+```
+
+Then, we define how to branch to identifiers, and how the branches using integers still match on named labels.
+
+```k
+    rule <k> br 0     ~> ( label _:Identifier [ TYPES ] { IS } VALSTACK' => label [ TYPES ] { IS } VALSTACK' ) ... </k>
+
+    rule <k> br ID:Identifier ~> label ID  [ TYPES ] { IS } VALSTACK' => IS ... </k>
+         <valstack> VALSTACK => #take(TYPES, VALSTACK) ++ VALSTACK' </valstack>
+    rule <k> br ID:Identifier ~> label [ TYPES ] { IS } VALSTACK' => br ID ... </k>
+    rule <k> br ID:Identifier ~> label ID' [ TYPES ] { IS } VALSTACK' => br ID ... </k>
+      requires ID =/=K ID'
```

Comment:
formatting here is a little weird

Reply:
Also, `br` is semantics, it feels weird to duplicate these here. Are we sure there isn't a way to desugar this to use the integer index? Otherwise, let's move back over to the semantics for now.

Reply:
Move backed to core semantics.